<!-- "map"->"odom"のtfを推定し発行、ブレる.  -->

<launch>
  <env name="ROSCONSOLE_CONFIG_FILE"
       value="$(find for_testing)/use_debug_level.conf"/>
    <!-- <include file="$(find chairbot_selfdrive)/launch/include/navsat_transform.launch"/> -->

    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_absolute" clear_params="true" output="screen">
      <param name="frequency" value="30"/>
      <param name="sensor_timeout" value="10.0"/>
      <param name="two_d_mode" value="true"/>

      <param name="map_frame" value="map"/>
      <param name="odom_frame" value="odom"/>
      <param name="base_link_frame" value="base_footprint"/>
      <param name="world_frame" value="map"/>
      <!-- <param name="world_frame" value="odom"/> -->

      <!-- Use this parameter to provide an offset to the transform generated by ekf_localization_node. This
           can be used for future dating the transform, which is required for interaction with some other
           packages. Defaults to 0.0 if unspecified. -->
      <param name="transform_time_offset" value="0.0"/>

      <param name="odom0" value="/mimamorukun/odom"/>
      <param name="odom1" value="odometry/gps"/>
      <param name="imu0" value="/imu/data"/>

      <rosparam param="odom0_config">[false, false, false,
                                      false, false, false,
                                      true,  false, false,
                                      false, false, true,
                                      false, false, false]</rosparam>
      <rosparam param="odom1_config">[true,  true,  false,
                                      false, false, false,
                                      false, false, false,
                                      false, false, false,
                                      false, false, false]</rosparam>
     <rosparam param="imu0_config">[false, false, false,
                                    false, false, false,
                                    false, false, false,
                                    false, false, true,
                                    false, false, false]</rosparam>

      <param name="odom0_differential" value="false"/>
      <param name="odom1_differential" value="false"/>
      <param name="imu0_differential" value="false"/>
      <!-- <param name="odom0_relative" value="false"/>
      <param name="odom1_relative" value="true"/>
      <param name="imu0_relative" value="false"/> -->
      <param name="imu0_remove_gravitational_acceleration" value="true"/>

      <param name="print_diagnostics" value="true"/>

      <param name="odom0_queue_size" value="100"/>
      <param name="odom1_queue_size" value="100"/>
      <param name="imu0_queue_size" value="100"/>

      <!-- If your data is subject to outliers, use these threshold settings, expressed as Mahalanobis distances, to control
           how far away from the current vehicle state a sensor measurement is permitted to be. Each defaults to
           numeric_limits<double>::max() if unspecified. -->
      <!-- <param name="odom1_pose_rejection_threshold" value="5"/>
      <param name="odom1_twist_rejection_threshold" value="1"/>
      <param name="imu0_twist_rejection_threshold" value="0.1"/> -->
      <!-- <param name="imu0_linear_acceleration_rejection_threshold" value="0.1"/> -->

      <!-- Debug settings. Not for the faint of heart. Outputs a ludicrous amount of information to the file
           specified by debug_out_file. I hope you like matrices! Defaults to false if unspecified. -->
      <param name="debug"           value="true"/>
      <!-- Defaults to "robot_localization_debug.txt" if unspecified. -->
      <param name="debug_out_file"  value="debug_ekf_localization.txt"/>

      <!-- The process noise covariance matrix can be difficult to tune, and can vary for each application, so it
           is exposed as a configuration parameter. The values are ordered as x, y, z, roll, pitch, yaw, vx, vy, vz,
           vroll, vpitch, vyaw, ax, ay, az. Defaults to the matrix below if unspecified. -->
      <rosparam param="process_noise_covariance">[0.05, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0.05, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0.06, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0.03, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0.03, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0.06, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0.025, 0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0.025, 0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0.04, 0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0.01, 0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.01, 0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.02, 0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.01, 0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.01, 0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.015]</rosparam>

      <!-- This represents the initial value for the state estimate error covariance matrix. Setting a diagonal value (a
           variance) to a large value will result in early measurements for that variable being accepted quickly. Users should
           take care not to use large values for variables that will not be measured directly. The values are ordered as x, y,
           z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. Defaults to the matrix below if unspecified. -->
           <rosparam param="initial_estimate_covariance">[1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam>


      <!-- Placeholder for output topic remapping -->
      <remap from="odometry/filtered" to="odometry/ekf_absolute"/>
    </node>


    <node pkg="raucha_utils" type="navsatfix_filter.py" name="navsatfix_filter">
      <remap from="navsatfix_in" to="navsat/fix"/>
      <remap from="navsatfix_out" to="navsat/filtered"/>
      <param name="cov_thresh" value="10000"/>
    </node>


    <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" respawn="true" output="screen">
      <param name="frequency" value="30"/>
      <param name="delay" value="0"/>
      <param name="magnetic_declination_radians" value="0"/>
      <param name="yaw_offset" value="0"/>
      <param name="zero_altitude" value="true"/>
      <param name="broadcast_utm_transform" value="true"/>
      <param name="publish_filtered_gps" value="true"/>
      <param name="use_odometry_yaw" value="true"/>
      <param name="wait_for_datum" value="false"/>
      <!-- <rosparam param="datum">[33.595755, 130.21930, 0.0, map, base_link]</rosparam> -->
      <remap from="/imu/data" to="/imu/data"/>
      <remap from="/odometry/filtered" to="/odometry/ekf_absolute"/>
      <!-- <remap from="/odometry/filtered" to="/mimamorukun/odom"/> -->
      <remap from="/gps/fix" to="/navsat/filtered"/>
      <!-- <remap from="/gps/fix" to="/navsat/fix"/> -->
      <!-- Placeholders for output remapping.
      <remap from="/odometry/gps" to=""/>
      <remap from="/gps/filtered" to=""/>
      -->
    </node>

    <!-- <node name="gps_conv" pkg="gps_common" type="utm_odometry_node">
      <remap from="odom" to="/gps_conv"/>
      <remap from="fix" to="/navsat/fix" />
      <param name="rot_covariance" value="99999" />
      <param name="frame_id" value="utm" />
      <param name="child_frame_id" value="gps_conv" />
    </node> -->
</launch>
